Warn if fifo tasks gets a thread limit

aconsroe-hpe opened the pull request:
When CHPL_TASKS=fifo and CHPL_RT_NUM_THREADS_PER_LOCALE is set in the
environment, there is a much higher chance we deadlock because we
might run out of threads to do work that would unblock blocked
threads.

This adds a warning in this situation and an accompanying setting
CHPL_RT_NUM_THREADS_PER_LOCALE_QUIET that can be set to suppress this warning

Add documentation on how to suppress the warning

Closes #18867

Passes paratest with CHPL_TASKS=fifo

ronawho commented on runtime/src/tasks/fifo/tasks-fifo.c:
```diff
@@ -310,6 +312,8 @@ void chpl_task_init(void) {
 
   initialized = true;
+
+  warn_threads_per_locale();
```
I would move the warning into `chpl_thread_init` where we call `chpl_task_getenvNumThreadsPerLocale`. Maybe something like:

```c
if ((lim = chpl_task_getenvNumThreadsPerLocale()) > 0) {
  chpl_task_warnNumThreadsPerLocale("Setting number of threads in CHPL_TASKS=fifo can lead to deadlock")
  maxThreads = lim;
}
```

And then` chpl_task_warnNumThreadsPerLocale` would exist as a sibling to `chpl_task_getenvNumThreadsPerLocale`. We could use this warning wrapper in `chpl_task_getenvNumThreadsPerLocale` itself if threads exceeds comm limit and in qthreads  `setupAvailableParallelism` where we currently warn (though for some reason not using `chpl_warning`) if `chpl_task_getenvNumThreadsPerLocale` is higher than the number of logical cores.

aconsroe-hpe replied:
If I move it into `chpl_thread_init`, what mechanism can I use to tell if CHPL_TASKS=fifo? I looked for a macro etc but didn't see anything

ronawho replied:
`CHPL_TASKS=fifo` requires `CHPL_THREADS=pthreads` and vice-versa. It used to be that we had different threading layers under fifo, but that's not true anymore and ideally we'd just merge them into a single layer, but it hasn't been worth the time to make that change.

ronawho replied:
And technically you could check if `CHPL_TASKS_FIFO` is defined (set by `chpl-env-gen.h`), but I don't think there's any reason to.

aconsroe-hpe replied:
Updated with that suggestion. Didn't touch qthreads though

ronawho commented on runtime/src/tasks/fifo/tasks-fifo.c:
```diff
@@ -993,3 +997,16 @@ uint32_t chpl_task_getNumThreads(void) {
+// Misc
+
+static void warn_threads_per_locale(void) {
+  if (getenv("CHPL_RT_NUM_THREADS_PER_LOCALE_QUIET")) return;
```
We have `chpl_env_rt_get_bool` wrappers I think'd we'd want to use (it does truthy/falsey checks so you could have `CHPL_RT_NUM_THREADS_PER_LOCALE_QUIET=false`). 

Not something you have to do, but if you were copying code from `chpl_task_getenvNumThreadsPerLocale`, I believe most of that code predates the wrappers. It seems like that code could be cleaned up by using the wrappers though (won't be as straightforward as just using the int parser since we have sentinel values, but may clean things up to use `chpl_env_rt_get` and then `chpl_env_str_to_int` when we need the int.)

`chpl_task_getEnvCallStackSize` also seems like it could be simplified with `chpl_env_str_to_size`, but I don't mean to add to what was supposed to be a simple PR, just noting is since I found it while looking at the other.

aconsroe-hpe replied:
Looked at using the wrappers in `chpl_task_getenvNumThreadsPerLocale` and there's enough different in error handling that I'm leaving it unless there's a major push to clean it up.

ronawho commented on runtime/src/chpl-tasks.c:
```diff
@@ -37,17 +37,29 @@
     return num;
 
-  if ((p = getenv("CHPL_RT_NUM_THREADS_PER_LOCALE")) != NULL) {
+  if ((p = chpl_env_rt_get("RT_NUM_THREADS_PER_LOCALE", NULL)) != NULL) {
```
Drop the `RT_` here I think

ronawho commented on runtime/src/chpl-tasks.c:
```diff
@@ -37,17 +37,29 @@
+
+  if (quiet == -1) {
+    quiet = chpl_env_rt_get_bool("NUM_THREADS_PER_LOCALE_QUIET", 0);
+  }
```
I don't think it's worth caching this. I'd probably just do 
```c
if (chpl_env_rt_get_bool("NUM_THREADS_PER_LOCALE_QUIET", false)) {
    chpl_warning(msg, 0, 0);
}
```

I'm not against caching it if you want to, I mostly just noticed it because of the mixed int/bool nature here (which is legal in C, but just looked a little odd to have something that's logically a bool with 3 possible values)

aconsroe-hpe replied:
Updated, caching was overboard

aconsroe-hpe commented:
Squashed and updated the PR text. Removed caching. Added docs.

I included the warning message text in the docs so that it might be found when googling

ronawho commented:
Looks good! Mark as closing https://github.com/chapel-lang/chapel/issues/18867 in the PR description?

aconsroe-hpe commented:
Looking for suggestions on how to handle some test failures, should we default to setting `CHPL_RT_NUM_THREADS_PER_LOCALE_QUIET=yes` or update the `.good` files to match?

The failing tests are:
  * domains/sungeun/assoc/stress.numthr
  * parallel/begin/dinan/mvm_coforall
  * parallel/coforall/bradc/manyThreads-inorder
  * parallel/taskPool/figueroa/ManyThreads
  * runtime/gbt/numThreadsSymbolicLogical
  * runtime/gbt/numThreadsSymbolicPhysical
From a spot check they are all just because the extra warning causes a diff failure

ronawho commented:
> domains/sungeun/assoc/stress.numthr
> parallel/begin/dinan/mvm_coforall
> parallel/coforall/bradc/manyThreads-inorder


These already have a prediff to quiet the qthreads warning. My thought would be to update qthreads warning to use your new quiet code, remove the prediff, and set the env var in the .execenv.

> parallel/taskPool/figueroa/ManyThreads

This test is checking that we warn if the number of threads is too large for the comm layer. I think we should update this test to also check that we get the fifo warning. Should be able to do that with a fifo specific .good file. 

> runtime/gbt/numThreadsSymbolicLogical
> runtime/gbt/numThreadsSymbolicPhysical

Let's add the QUIET env var to the .execenv.

ronawho commented:
> This test is checking that we warn if the number of threads is too large for the comm layer.

Might also be nice to use your new quiet warning here too. i.e. change the following places to call `chpl_task_warnNumThreadsPerLocale`:

https://github.com/chapel-lang/chapel/blob/1ea812b3af1d56c3d00fedda2e276f2937c61da1/runtime/src/chpl-tasks.c#L78

https://github.com/chapel-lang/chapel/blob/1ea812b3af1d56c3d00fedda2e276f2937c61da1/runtime/src/tasks/qthreads/tasks-qthreads.c#L450-L454

(not required, but if you're up for changing the qthreads code, I think this is the last place we'd want to use the quiet wrapper)

ronawho commented on doc/rst/usingchapel/tasks.rst:
```diff
@@ -300,6 +300,14 @@ CHPL_TASKS == fifo
   can result in deadlock if there are not enough threads to implement
   all of the required tasks.
 
+  When ``CHPL_RT_NUM_THREADS_PER_LOCALE`` is set, a warning is issued like:
```
"When ``CHPL_RT_NUM_THREADS_PER_LOCALE`` is set" -> "When ``CHPL_RT_NUM_THREADS_PER_LOCALE`` is set with fifo tasking"?

ronawho replied:
Nvm, I now see this is under a fifo section.

aconsroe-hpe commented:
I'ved updated the failing tests. I did replace that spot in qthreads and I did end up adding support for `.tasks-fifo.good` file as I didn't want to squash the comm-gasnet error as well. And that test is the one place where we are checking if our warning is firing.

ronawho commented on doc/rst/developer/bestPractices/TestSystem.rst:
```diff
@@ -346,8 +346,13 @@ explicitly in the ``.compopts`` or ``.execopts`` file for the test.
   will then apply for CHPL_COMM != none)
 - ``.no-local.good``: used with ``--no-local`` testing
 - ``.lm-numa.good``: used with CHPL_LOCALE_MODEL=numa
+- ``.na-ofi.good``: used with CHPL_NETWORK_ATOMICS=ofi
```
In practice this is always `.na-none.good`. i.e. the default tests with network atomics, and this overwrites to check without network atomics. Comm counts / test output should be the same for all comm layers that support network atomics, so I wouldn't expect to ever use na-ofi or na-ugni or something.

ronawho commented on runtime/src/tasks/fifo/tasks-fifo.c:
```diff
@@ -906,7 +906,7 @@ static void maybe_add_thread(void) {
                 ", but unable to create more than %d threads",
                 num_threads);
-      chpl_warning(msg, 0, 0);
+      chpl_task_warnNumThreadsPerLocale(msg);
```
I don't think we want to use the wrapper here. I think the wrapper makes sense to quiet warnings we issues when the user has set `CHPL_RT_NUM_THREADS_PER_LOCALE`, but that's not the case here.

aconsroe-hpe replied:
Gotcha that makes sense

aconsroe-hpe commented:
Not too nervous. The good file seemed like a better solution for ourselves in the future
